examples/remoting/stateful_remote_spawn.cpp

```diff
@@ -0,0 +1,176 @@
+                                    caf::result<int32_t>(caf::get_atom)>;
+};
+using aggregator = caf::typed_actor<aggregator_trait>;
+
+CAF_BEGIN_TYPE_ID_BLOCK(remote_spawn, first_custom_type_id)
+
+  CAF_ADD_TYPE_ID(remote_spawn, (aggregator))
+
+CAF_END_TYPE_ID_BLOCK(remote_spawn)
+
+using std::string;
+
+using namespace caf;
+
+struct aggregator_state {
+  aggregator_state(int32_t init) : value(init) {
+  }
+  aggregator::behavior_type make_behavior() {
+    return {
+      [this](add_atom, int32_t a) { value += a; },
+      [this](get_atom) -> result<int32_t> { return value; },
+    };
+  }
+  int32_t value;
+};
+
+// removes leading and trailing whitespaces
+string trim(string s) {
+  auto not_space = [](char c) { return isspace(c) == 0; };
+  // trim left
+  s.erase(s.begin(), find_if(s.begin(), s.end(), not_space));
+  // trim right
+  s.erase(find_if(s.rbegin(), s.rend(), not_space).base(), s.end());
+  return s;
+}
+
+// implements our main loop for reading user input
+void client_repl(actor_system& sys, aggregator hdl) {
+  auto usage = [&sys] {
+    sys.println("Usage:");
+    sys.println("  quit                  : terminate program");
+    sys.println("  add <x>               : adds x to remote aggregate");
+    sys.println("  get                   : prints the aggregated value");
+    sys.println("");
+  };
+  usage();
+  scoped_actor self{sys};
+  self->link_to(hdl);
+  // read next line, split it, and evaluate user input
+  string line;
+  while (std::getline(std::cin, line)) {
+    line = trim(std::move(line));
+    if (line == "quit")
+      return;
+    std::vector<string> words;
+    split(words, line, is_any_of(" "), token_compress_on);
+    if (words.size() > 2) {
+      usage();
+      continue;
+    }
+    using namespace std::chrono_literals;
+    auto to_int32_t = [](const string& str) -> std::optional<int32_t> {
+      char* end = nullptr;
+      auto res = strtol(str.c_str(), &end, 10);
+      if (end == str.c_str() + str.size())
+        return static_cast<int32_t>(res);
+      return std::nullopt;
+    };
+    if (words[0] == "add" && words.size() == 2) {
+      auto x = to_int32_t(words[1]);
+      if (!x) {
+        usage();
+        continue;
+      }
+      self->mail(add_atom_v, *x).send(hdl);
+    } else if (words[0] == "get") {
+      auto value = self->mail(get_atom_v).request(hdl, 1s).receive();
+      if (value) {
+        sys.println("Aggregated: {}", *value);
+      } else {
+        sys.println("Error fetching value: {}", value.error());
+      }
+    } else {
+      usage();
+      continue;
+    }
+  }
+}
+
+constexpr uint16_t default_port = 0;
+constexpr std::string_view default_host = "localhost";
+constexpr bool default_server_mode = false;
+
+struct config : actor_system_config {
+  config() {
+    // Constructor parameters are listed after naming the actor.
+    add_actor_type<aggregator_state, int32_t&>("aggregator");
```

Comment:
I dislike the interface. One of the main drivers of getting rid of class-based spawn was the asymmetry between class-based and function-based spawn.

If we have a local spawn, we have:

```c++
  sys.spawn(my_fun, 1, 2, 3); // function
  sys.spawn(actor_from_state<my_state>, 1, 2, 3); // state class
```

The interface for remote spawn should be analogous:

```c++
  add_actor_type("calculator", my_fun);
  add_actor_type("calculator", actor_from_state<my_state>);
```

We can't infer the constructor arguments (at least not in C++17), so we probably have to pass another token for that, but I would prefer something like this over the current approach:

```c++
  add_actor_type("calculator", my_fun);
  add_actor_type("calculator", actor_from_state<my_state1>); // no extra ctor args
  add_actor_type("calculator", actor_from_state<my_state2>, type_list_v<int32_t>);  
```

And yes, we should definitely get rid of the references.

(note: `type_list_v` does not exist yet 🙂)

Reply:
I like the proposed idea, if you've meant to say `calculator1` and `calculator2` in the last code snippet (I hope so... :smile: ).
Also, I've supported multiple constructors:
```
add_actor_type("calculator", actor_from_state<my_state2>, type_list_v<int32_t>, 
               type_list_v<std::string>, type_list_v<double>);
```

Reply:
Btw, great idea on supporting multiple constructors! 👍